File: web/src/game/Deck.ts
import { Card, CardColor } from "./Card";
import { Hand } from "./Hand";

export class Deck {
  private cards: Card[] = [];

  constructor() {
    for (const color of Object.values(CardColor)) {
      for (let value = 1; value <= 13; value++) {
        this.cards.push(new Card(value, color));
      }
    }
  }

  shuffle(): void {
    for (let i = this.cards.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [this.cards[i], this.cards[j]] = [this.cards[j], this.cards[i]];
    }
  }

  size(): number {
    return this.cards.length;
  }

  drawCard(): Card | null {
    if (this.size() === 0) return null;
    return this.cards.shift() ?? null;
  }

  drawCards(num: number): Hand | null {
    if (this.cards.length < num) return null;
    const drawn: Card[] = [];
    for (let i = 0; i < num; i++) {
      drawn.push(this.cards.shift()!);
    }
    return new Hand(drawn);
  }
}

File: shared/src/commonMain/kotlin/com/example/tp4/Deck.kt
package com.example.tp4

class Deck {

    private val cards: MutableList<Card>

    init {
        cards = mutableListOf()

        for (color in Card.Color.entries) {
            for (value in 1..13) {
                cards.add(Card(value, color))
            }
        }
    }

    fun Shuffle() {
        cards.shuffle()
    }

    fun size(): Int = cards.size

    fun drawCard(): Card? {
        if (size() == 0) return null
        return cards.removeAt(0)
    }

    fun drawCards(num: Int): Hand? {
        if (cards.size < num) {
            return null
        }

        val drawnCards = mutableListOf<Card>()
        repeat(num) {
            drawnCards.add(cards.removeAt(0))
        }

        return Hand(drawnCards)
    }

    override fun toString(): String {
        val sb = StringBuilder()
        for (i in cards.indices) {
            sb.append(cards[i]).append(" ")

            if ((i + 1) % 13 == 0) {
                sb.append("\n")
            }
        }
        return sb.toString()
    }
}
